x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,429 @@
+			return types.ErrSwapOutFailed.Wrapf("the successor sp(ID: %d) can not be the primary sp of gvg(%s).", successorSPID, gvg.String())
+		}
+		secondarySPFound := false
+		for _, spID := range gvg.SecondarySpIds {
+			if spID == successorSPID {
+				return types.ErrSwapOutFailed.Wrapf("the successor sp(ID: %d) can not be one of the secondary sp of gvg(%s).", successorSPID, gvg.String())
+			}
+			if spID == secondarySPID {
+				secondarySPFound = true
+			}
+		}
+		if !secondarySPFound {
+			return types.ErrSwapOutFailed.Wrapf("The sp(ID: %d) that needs swap out is not one of the secondary sps of gvg gvg(%s).", secondarySPID, gvg.String())
+		}
+		// settlement
+		err := k.SettleAndDistributeGVG(ctx, gvg)
+		if err != nil {
+			return types.ErrSwapOutFailed.Wrapf("fail to settle GVG %d", gvgID)
+		}
+
+		for i, spID := range gvg.SecondarySpIds {
+			if spID == secondarySPID {
+				gvg.SecondarySpIds[i] = successorSPID
+				origin := k.MustGetGVGStatisticsWithinSP(ctx, secondarySPID)
+				successor := k.MustGetGVGStatisticsWithinSP(ctx, successorSPID)
+				origin.SecondaryCount--
+				successor.SecondaryCount++
+				k.SetGVGStatisticsWithSP(ctx, origin)
+				k.SetGVGStatisticsWithSP(ctx, successor)
+				k.SetGVG(ctx, gvg)
+				break
+			}
+		}
+	}
+	return nil
+}
+
+func (k Keeper) GetOrCreateGVGStatisticsWithinSP(ctx sdk.Context, spID uint32) *types.GVGStatisticsWithinSP {
+	store := ctx.KVStore(k.storeKey)
+
+	ret := &types.GVGStatisticsWithinSP{
+		StorageProviderId: spID,
+		SecondaryCount:    0,
+	}
+	bz := store.Get(types.GetGVGStatisticsWithinSPKey(spID))
+	if bz == nil {
+		return ret
+	}
+
+	k.cdc.MustUnmarshal(bz, ret)
+	return ret
+}
+
+func (k Keeper) GetGVGStatisticsWithinSP(ctx sdk.Context, spID uint32) (*types.GVGStatisticsWithinSP, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetGVGStatisticsWithinSPKey(spID))
+	if bz == nil {
+		return nil, false
+	}
+
+	var ret types.GVGStatisticsWithinSP
+	k.cdc.MustUnmarshal(bz, &ret)
+	return &ret, true
+}
+
+func (k Keeper) MustGetGVGStatisticsWithinSP(ctx sdk.Context, spID uint32) *types.GVGStatisticsWithinSP {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetGVGStatisticsWithinSPKey(spID))
+	if bz == nil {
+		panic("must get gvg statistics within sp")
+	}
+
+	var ret types.GVGStatisticsWithinSP
+	k.cdc.MustUnmarshal(bz, &ret)
+	return &ret
+}
+
+func (k Keeper) SetGVGStatisticsWithSP(ctx sdk.Context, gvgsStatisticsWithinSP *types.GVGStatisticsWithinSP) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(gvgsStatisticsWithinSP)
+
+	store.Set(types.GetGVGStatisticsWithinSPKey(gvgsStatisticsWithinSP.StorageProviderId), bz)
+}
+
+func (k Keeper) BatchSetGVGStatisticsWithinSP(ctx sdk.Context, gvgsStatisticsWithinSP []*types.GVGStatisticsWithinSP) {
+	for _, g := range gvgsStatisticsWithinSP {
+		k.SetGVGStatisticsWithSP(ctx, g)
+	}
+}
+
+func (k Keeper) IsStorageProviderCanExit(ctx sdk.Context, spID uint32) error {
```

Comment:
`IsStorageProviderCanExit` ==> `StorageProviderExitable`

Reply:
Exitable?  seems this is not a valid word.